app/src/utils/chainflip.ts

```diff
@@ -0,0 +1,357 @@
+import type {
+  AssetData,
+  ChainData,
+  DCAQuote,
+  DepositAddressResponseV2,
+  RegularQuote,
+  SwapSDK,
+  SwapStatusResponseV2,
+} from '@chainflip/sdk/swap'
+import {
+  type Chain,
+  chainflipRoutes,
+  EthereumTokens,
+  MainnetRegistry,
+  PolkadotTokens,
+  type Token,
+} from '@velocitylabs-org/turtle-registry'
+import type { FeeDetailType, StoredTransfer } from '@/models/transfer'
+import { useChainflipStore } from '@/store/chainflipStore'
+import { getChainSpecificAddress } from './address'
+
+/** TYPES */
+export type AssetSymbol = 'DOT' | 'USDC' | 'USDT' | 'ETH' | 'FLIP' | 'BTC' | 'SOL'
+
+export type ChainflipChain = 'Ethereum' | 'Polkadot' | 'Assethub' | 'Arbitrum' | 'Bitcoin' | 'Solana'
+
+export type ChainflipFeeType = 'NETWORK' | 'INGRESS' | 'EGRESS' | 'BROKER' | 'BOOST' | 'REFUND'
+
+export type ChainflipQuote = RegularQuote | DCAQuote
+
+type ChainflipError = {
+  response?: {
+    data?: {
+      message?: string
+    }
+  }
+}
+
+export enum ChainflipRefundStatus {
+  Refunded = 'refunded',
+  Partially = 'partially refunded',
+}
+
+/** ROUTES HELPERS */
+
+/** Returns all Chainflip allowed source chains for a swap. */
+export const getChainflipSwapSourceChains = (): Chain[] => {
+  return chainflipRoutes.map(route => route.from)
+}
+
+/** Returns all Chainflip allowed source tokens for a swap. */
+export const getChainflipSwapSourceTokens = (sourceChain: Chain): Token[] => {
+  if (!sourceChain) return []
+
+  const tokensSet = new Set<Token>()
+
+  chainflipRoutes
+    .filter(route => route.from.chainId === sourceChain.chainId)
+    .map(route => route.pairs.map(([token, _]) => tokensSet.add(token)))
+
+  return Array.from(tokensSet)
+}
+
+/** Returns all Chainflip allowed destination chains for a swap. */
+export const getChainflipSwapDestChains = (sourceChain: Chain, sourceToken: Token): Chain[] => {
+  if (!sourceChain || !sourceToken) return []
+
+  const chainsSet = new Set<Chain>()
+
+  chainflipRoutes.forEach(route => {
+    if (route.from.chainId === sourceChain.chainId && route.pairs.some(([token, _]) => token.id === sourceToken.id)) {
+      chainsSet.add(route.to)
+    }
+  })
+
+  return Array.from(chainsSet)
+}
+
+/** Returns all Chainflip allowed destination tokens for a swap. */
+export const getChainflipSwapDestTokens = (
+  sourceChain: Chain,
+  sourceToken: Token,
+  destinationChain: Chain | null,
+): Token[] => {
+  if (!sourceChain || !sourceToken || !destinationChain) return []
+
+  const tokensSet = new Set<Token>()
+  const route = chainflipRoutes.find(
+    route => route.from.chainId === sourceChain.chainId && route.to.chainId === destinationChain.chainId,
+  )
+  if (!route) return []
+
+  route.pairs.forEach(([srcT, destT]) => srcT.id === sourceToken.id && tokensSet.add(destT))
+
+  return Array.from(tokensSet)
```

Comment:
I would prefer a more functional structure to this. Plus expressions like "srcT.id === sourceToken.id && tokensSet.add(destT)` are borderline magic, it's not immediately understandable code.

```suggestion
 const desTokens = chainflipRoutes.find(route => 
   route.from.chainId === sourceChain.chainId && route.to.chainId === destinationChain.chainId,
  )?.pairs
  .filter(([srcT, _]) => srcT.id === sourceToken.id)
  .map(([_, destT]) => destT)


// return uniques
return [...new Set(destTokens)];
```

Reply:
The list of routes and pairs we are handling is minimal so performance concerns is not a real issue in practice. In cases like that I prefer to prioritise code readability but that's also not a clear cut here either. I let you decide on this.

Reply:
agree with @NunoAlexandre on this one, the `&& .add` is javascript being a very permissive mom